### ISSUE TITLE
Non flagged blown up WEIGHT_SPECTRUM values

### ISSUE THREAD

**Issue by ajstewart** [lofar-astron/PiLL#10]
Not a PiLL fault as such but rather a question on the output (mainly to @revoltek). 

Inspecting the logs from a PiLL run I spotted:
```
2017-03-15 17:36:09 DEBUG   node.pi3.executable_args.L229405_SAP000_SB114_uv.postsagecal.HBAflag.dppp.flagtar: /home/stewart/pi3raid/ncp-2017/PiLL-target/scripts/flag_weight_to_zero.py stdout: Sum weights before: 25468267007049728.000000
Resetting 5423392.000000 values
Sum weights after: 1646.375244

2017-03-15 17:36:09 DEBUG   node.pi3.executable_args.L229405_SAP000_SB118_uv.postsagecal.HBAflag.dppp.flagtar: /home/stewart/pi3raid/ncp-2017/PiLL-target/scripts/flag_weight_to_zero.py stdout: Sum weights before: 35148537029197824.000000
Resetting 5655056.000000 values
Sum weights after: 1663.950317

2017-03-15 17:36:09 DEBUG   genericpipeline.executable_args: Results for job 104 submitted by ('163.1.137.43', 53236)
2017-03-15 17:36:09 INFO    node.pi3.executable_args: Total time 4.0173s; user time: 0.3640s; system time: 1.4680s
2017-03-15 17:36:09 DEBUG   node.pi3.executable_args: Start time was 1489599365.4533s; end time was 1489599369.4710s
2017-03-15 17:36:09 DEBUG   node.pi3.executable_args.L229405_SAP000_SB120_uv.postsagecal.HBAflag.dppp.flagtar: /home/stewart/pi3raid/ncp-2017/PiLL-target/scripts/flag_weight_to_zero.py stdout: Sum weights before: nan
Resetting 22226428.000000 values
Sum weights after: 140011917524364720962793961271148412928.000000
```
So something has caused a non-flagged weight spectrum column in SB120 to blow up. I've checked every other flag weight output and they are similar to the other two above - though are these sums still too high? They are all in the range of a few hundred - 3000. Even including the ones with ```nan```, this is the only one that's huge after.

I narrowed it down to only two rows in the MS (below is the weight_spectrum column):
```
In [5]: print b[1826881],b[1827409]
[[  3.47274213e+36   3.28606378e+36   6.02758099e+36   5.70356633e+36]
 [  3.64336949e+36   3.53033118e+36   6.29758867e+36   6.10220188e+36]
 [  3.61579682e+36   3.48250748e+36   6.32384108e+36   6.09072521e+36]
 [  4.30689520e+36   4.08962623e+36   7.37698619e+36   7.00484136e+36]] 
[[  3.10430233e+36   2.93742501e+36   5.38808803e+36   5.09844159e+36]
 [  2.85121019e+36   2.76274529e+36   4.92833841e+36   4.77542647e+36]
 [  2.44136924e+36   2.35138491e+36   4.26982974e+36   4.11245155e+36]
 [  2.76740581e+36   2.62777980e+36   4.74010244e+36   4.50094652e+36]]
```
and double checked that they are ok in my archive before PiLL fetches them (which they are). Both rows are FLAG false.

I don't understand completely why this happens, I know something to do with transferring solutions and changing flags? Is it a fault in DPPP? I'm wondering if there is any further filter tool to flag such high weight_spectrum values that could be used to avoid letting such high values through?

**Comment by revoltek** [comment]
Are those autocorrelations? This problem of the weight is a damn pain. The new version of NDPPP should fix (part of) it. The problem is also more complicated because of the re-normalisation done by Dysco, if you have crazy high values, dysco screws up all the rest.

**Comment by adrabent** [comment]
In the genericpipeline version Dysco is not yet implemented.

**Comment by revoltek** [comment]
ok, then the re-setting of the weights is not necessary. However, it would be very helpful if someone would go through the life of the WEIGHT_SPECTRUM column along all pipelines to check whether everything make sense :)

**Comment by ajstewart** [comment]
> Are those autocorrelations? 

They are not it turns out:
```
In [3]: ant1=a.getcell("ANTENNA1", 1826881)

In [4]: ant2=a.getcell("ANTENNA2", 1826881)

In [5]: print ant1, ant2
0 1
```
It's the same for the other row.

> ok, then the re-setting of the weights is not necessary.

So the flag weights script is only applicable if going on to use dysco? If it doesn't hurt I guess it's worth leaving in to future proof it?

@adrabent Do you reckon it would be possible to include some sort of monitor script for the weight spectrum column? As then it could be inserted at every necessary step without hopefully very much effort... though perhaps easier said!

**Comment by adrabent** [comment]
The main question is: What should such a script capture? The entire column? Average values for every baseline? Should there be plots?

**Comment by ajstewart** [comment]
> The main question is: What should such a script capture? The entire column? Average values for every baseline? Should there be plots?

Yes I was just thinking this myself, the sum or average of every baseline was also my go to. I wonder if a histogram would be useful as an overview, as of course it would be a lot of information to convey, though extreme values might ruin this.

Maybe the simplest thing possible for the pipeline is better - only record the sum/average for each baseline in a easily readable format for each measurement set where necessary (I guess after each dppp action?), allowing these to be taken and analysed offline.

@revoltek - do you already have a metric of what to look for as regards the weights?

**Comment by revoltek** [comment]
this can be a starting point: https://github.com/revoltek/scripts/blob/master/plot_weights.py

**Comment by ajstewart** [comment]
The plots that the above script generates are actually quite useful to monitor the weights. The dataVSchan plot can probably be skipped in this case. Though the x axis range on the weightVSchan needs to be pushed back slightly to see the channel 0 values.

Reading back over this I guess there are two things that come from it:

1. A monitor for the weights during the pipeline to be able to check that they are sensible throughout the calibration procedure (at least for now for commissioning purposes). I think the plots above are ok for this, as like other plots during the process, these can be looked at after completion with any issues acted upon after.

2. The specific problem I ran into was due to the inclusion of a problematic SB that had somehow gained extremely high weight values that were not flagged (if the monitoring is implemented then this should help pinpoint exactly when this happened). Hence these rows would need flagging or simply for this sub band to be excluded.

   I'll probably write a small script similar to flag_weight_to_zero.py to simply mark any SBs with extremely high non-flagged weight values compared to the rest, then either flag those specific rows or remove the SB all together. If it is decided to, this is something that could be implemented quite easily in PiLL I assume, and need only be run once at the end before the concatenation.

   As this seems a bit odd that weights have blown up that are not flagged, it probably warrants a bit of further investigation to attempt to see exactly why this happened. But perhaps this is important enough to safeguard against regardless? While it doesn't cause any errors as PiLL is now, it becomes very important when the self-cal pipeline is implemented as it causes certain images to fail with certain weightings (how I spotted it).

Thankfully this problem seems rare and also hopefully to be (maybe) fixed in the future with a newer NDPPP. Does anyone know if this is specific to LBA, or if (pre)factor also has had to deal with blown up weight values?

**Comment by revoltek** [comment]
New NDPPP should have weights of flagged data set to NaN, this should fix dysco and other related problems. It's weird that weights are not smooth across the band/time of the observations, they need to be. If they are not we need to track down what happen. Note that my pipeline update the weights in the timesplit part. This is done to compensate the erroneous weights given by the correlator and it is done by NDPPP using the autocorrelation values (reason why I carry them up until that time).

**Comment by ajstewart** [comment]
@revoltek Thanks for clarifying that. Yes I'm keen to try and figure out why this non-flagged weight spectrum value blew up. It could be specific to this one sub band in this dataset and I never see it again with any others, but would still be worthwhile knowing what happened.
